readme.txt
=== Style Cleaner for Elementor ===
Contributors: jflagarde
Donate link: https://jflagarde.com/
Tags: elementor, template, blocks, style, typography, colours
Requires at least: 5.3.2
Tested up to: 5.3.2
Stable tag: 0.0.2
License: GPL-3.0 or later
License URI: http://www.gnu.org/licenses/gpl-3.0.txt

Remove all the hard coded styling of all the elements on a page.

== Description ==

Remove all the hard coded styling of all the elements on a page.

== Installation ==

1. Upload the content of `jfl-style-cleaner.zip` to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Edit a page with Elementor
1. Right click on an element and use one of the added actions

== Changelog ==

0.0.2
- Fix: Downgraded the required PHP version to 7.0
- Tweak: Installation and usage instructions

= 0.0.1 =
* First proof of concept.

changelog.txt
== Changelog ==

31/01/2020
0.0.2
- Fix: Downgraded the required PHP version to 7.0
- Update: Installation and usage instructions

30/01/2020
0.0.1
- First proof of concept.

assets/js/jfl-style-cleaner.js
// When the Elementor editor is initialized
jQuery(window).on('elementor:init', function () {

    // Add actions to the contextual menus
    elementor.hooks.addFilter('elements/widget/contextMenuGroups', function (groups, widget) {
        return jflAddCleanerActions(groups, widget);
    });
    elementor.hooks.addFilter('elements/column/contextMenuGroups', function (groups, column) {
        return jflAddCleanerActions(groups, column);
    });
    elementor.hooks.addFilter('elements/section/contextMenuGroups', function (groups, section) {
        return jflAddCleanerActions(groups, section);
    });

});

/*
 * This function adds different actions to the contextual menu
 */
function jflAddCleanerActions(groups, element) {

    // Add actions to each contextual group
    jQuery.each(groups, function (index, value) {

        // Add actions to the general group
        if (value.name == 'general') {
            groups[index].actions.push(
                {
                    name: 'jfl-inspect-selected-element',
                    title: 'Inspect Selected Element',
                    callback: function () {
                        jflInspectSelectedElement(element);
                        return;
                    }
                }
            );
        }

        // Add actions to the clipboard group
        if (value.name == 'clipboard') {
            groups[index].actions.push(
                {
                    name: 'jfl-clear-typo-colours',
                    title: 'Clear Typo and Colours',
                    callback: function () {
                        jflClearTypoColours();
                        return;
                    }
                },
                {
                    name: 'jfl-reset-all-styles',
                    title: 'Reset All Styles',
                    callback: function () {
                        jflResetAllStyles();
                        return;
                    }
                }
            );
        }

        // Add actions to the save group
        if (value.name == 'save') {
            groups[index].actions.push(
                {
                    name: 'jfl-set-navigator-title',
                    title: 'Set Navigator Title',
                    callback: function () {
                        jflSetNavigatorTitle(element);
                        return;
                    }
                }
            );
        }

    });

    // Return the groups for the filter
    return groups;
}

/*
 * This function will show the element in the console to inspect it
 */
function jflInspectSelectedElement(element) {

    // Log the element in the console
    console.log('element', element);

    // return nothing for the callback
    return;
}

/*
 * This function will return all the containers on a page
 */
function jflGetAllContainers() {

    // Get all the containers of the document
    var allContainers = Object.values(elementor.getPreviewView()._getNestedViews()).map((view) => {

        // Remove the empty views
        if (view.el.className != 'elementor-empty-view' && !view.isDestroyed) {

            var container = view.getContainer();
            container.view.allowRender = false;
            return container;

        } else {

            // The view contained nothing
            return;

        }

    });

    // Clean-up the array of containers
    var filteredContainers = allContainers.filter((container) => {
        return container;
    });
    filteredContainers.reverse();

    // return the containners
    return filteredContainers;
}

/*
 * This function will remove all the hard coded typography and colours of all the elements on a page
 */
function jflClearTypoColours() {

    /*
    // List all the controls that we want to be resetted
    const controlsToReset = [
        'background_color',
        'border_color',
        'heading_color',
        'color_text',
        'typography_typography',
        'title_color',
        'title_typography_typography',
        'description_color',
        'description_typography_typography',
        'primary_color',
        'secondary_color',
        'content_content_color',
        'content_typography_typography',
        'name_text_color',
        'name_typography_typography'
    ];
    */

    // Get all the containers of the document
    var filteredContainers = jflGetAllContainers();

    // For each container
    filteredContainers.forEach((container) => {

        // Get the controls of the container
        const controls = container.settings.controls;
        const defaultValues = {};

        // Stop the rendering to avoid container issues
        container.view.allowRender = false;

        // For each control of the container
        Object.entries(controls).forEach(([controlName, control]) => {

            // From the Elementor code, not too sure what it does. Probably check to be sure the element has controls.
            if (!container.view.isStyleTransferControl(control)) {
                return;
            }

            /*
            // Check if the control is in our list we have defined manually above
            if ( controlsToReset.includes(controlName) ) {
                defaultValues[ controlName ] = control.default;
            }
            */

            // Reset all the settings ending with either color or typography, the method could be combined with the manual list if needed
            if (controlName.endsWith("color") || controlName.endsWith("typography")) {
                defaultValues[controlName] = control.default;
            }

        });

        // Reset the selected settings to their default values
        $e.run('document/elements/settings', {
            container,
            settings: defaultValues,
            options: {
                external: true,
            },
        });

        // Allow the container to be rendered
        container.view.allowRender = true;

    });

    // Render all the page
    elementor.getPreviewView().render();

    // return nothing for the callback
    return;
}

/*
 * This function will remove all the hard coded styling of all the elements on a page
 */
function jflResetAllStyles() {

    // Get all the containers of the document
    var filteredContainers = jflGetAllContainers();

    // Remove the styling of all containers
    $e.run('document/elements/reset-style', { containers: filteredContainers });

    // Render the document
    filteredContainers.forEach((container) => {
        container.view.allowRender = true;
    });
    elementor.getPreviewView().render();

    // return nothing for the callback
    return;
}

/*
 * This function will set the Navigator element title from the widget content
 */
function jflSetNavigatorTitle(element) {

    // Get the settings of the element
    const settingsModel = element.model.get('settings')

    // Find the best title from the attributes
    if (settingsModel.attributes.title) {
        newTitle = settingsModel.attributes.title;
    } else if (settingsModel.attributes.title_text) {
        newTitle = settingsModel.attributes.title_text;
    } else {
        newTitle = settingsModel.get('_title');
    }

    // Set the navigator title
    settingsModel.set('_title', newTitle);

    // return nothing for the callback
    return;
}

jfl-style-cleaner.php
<?php
/**
 * Style Cleaner for Elementor
 *
 * Remove all the hard coded styling of all the elements on a page.
 *
 * @copyright Copyright (C) 2020, JFLagarde - Consultant et stratège web - jflagarde@jflagarde.com
 * @license   http://www.gnu.org/licenses/gpl-3.0.html GNU General Public License, version 3 or higher
 *
 * @wordpress-plugin
 * Plugin Name: Style Cleaner for Elementor
 * Plugin URI:  https://jflagarde.com/
 * Description: Remove all the hard coded styling of all the elements on a page.
 * Version:     0.0.2
 * Author:      Jean-François Lagarde <jflagarde@jflagarde.com>
 * Author URI:  https://jflagarde.com/
 * Text Domain: jfl-style-cleaner
 * Domain Path: /languages
 * License:     GPL-3.0+
 * License URI: http://www.gnu.org/licenses/gpl-3.0.txt
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Style Cleaner for Elementor incorporates code from:
 * - Dynamic Content for Elementor, Copyright (c) 2018-2020 Ovation S.r.l., License: GPL-3.0+, https://www.dynamic.ooo/
 */

if ( ! defined( 'ABSPATH' ) ) {
    exit; // Exit if accessed directly.
}

/**
 * Main Style Cleaner for Elementor Class
 *
 * The main class that initiates and runs the plugin.
 *
 * @since 0.0.1
 */
final class Elementor_Test_Extension {

    /**
     * Plugin Version
     *
     * @since 0.0.1
     *
     * @var string The plugin version.
     */
    const VERSION = '0.0.2';

    /**
     * Minimum Elementor Version
     *
     * @since 0.0.1
     *
     * @var string Minimum Elementor version required to run the plugin.
     */
    const MINIMUM_ELEMENTOR_VERSION = '2.8.5';

    /**
     * Minimum PHP Version
     *
     * @since 0.0.1
     *
     * @var string Minimum PHP version required to run the plugin.
     */
    const MINIMUM_PHP_VERSION = '7.0';

    /**
     * Instance
     *
     * @since 0.0.1
     *
     * @access private
     * @static
     *
     * @var Elementor_Test_Extension The single instance of the class.
     */
    private static $_instance = null;

    /**
     * Instance
     *
     * Ensures only one instance of the class is loaded or can be loaded.
     *
     * @since 0.0.1
     *
     * @access public
     * @static
     *
     * @return Elementor_Test_Extension An instance of the class.
     */
    public static function instance() {

        if ( is_null( self::$_instance ) ) {
            self::$_instance = new self();
        }
        return self::$_instance;

    }

    /**
     * Constructor
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function __construct() {

        add_action( 'init', [ $this, 'i18n' ] );
        add_action( 'plugins_loaded', [ $this, 'init' ] );

    }

    /**
     * Load Textdomain
     *
     * Load plugin localization files.
     *
     * Fired by `init` action hook.
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function i18n() {

        load_plugin_textdomain( 'jfl-style-cleaner' );

    }

    /**
     * Initialize the plugin
     *
     * Load the plugin only after Elementor (and other plugins) are loaded.
     * Checks for basic plugin requirements, if one check fail don't continue,
     * if all check have passed load the files required to run the plugin.
     *
     * Fired by `plugins_loaded` action hook.
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function init() {

        // Check if Elementor installed and activated
        if ( ! did_action( 'elementor/loaded' ) ) {
            add_action( 'admin_notices', [ $this, 'admin_notice_missing_main_plugin' ] );
            return;
        }

        // Check for required Elementor version
        if ( ! version_compare( ELEMENTOR_VERSION, self::MINIMUM_ELEMENTOR_VERSION, '>=' ) ) {
            add_action( 'admin_notices', [ $this, 'admin_notice_minimum_elementor_version' ] );
            return;
        }

        // Check for required PHP version
        if ( version_compare( PHP_VERSION, self::MINIMUM_PHP_VERSION, '<' ) ) {
            add_action( 'admin_notices', [ $this, 'admin_notice_minimum_php_version' ] );
            return;
        }

        // Enqueue Editor Scripts
        add_action( 'elementor/editor/before_enqueue_scripts', [ $this, 'enqueue_editor_scripts' ] );
    }

    /**
     * Admin notice
     *
     * Warning when the site doesn't have Elementor installed or activated.
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function admin_notice_missing_main_plugin() {

        if ( isset( $_GET['activate'] ) ) unset( $_GET['activate'] );

        $message = sprintf(
            /* translators: 1: Plugin name 2: Elementor */
            esc_html__( '"%1$s" requires "%2$s" to be installed and activated.', 'jfl-style-cleaner' ),
            '<strong>' . esc_html__( 'Style Cleaner for Elementor', 'jfl-style-cleaner' ) . '</strong>',
            '<strong>' . esc_html__( 'Elementor', 'jfl-style-cleaner' ) . '</strong>'
        );

        printf( '<div class="notice notice-warning is-dismissible"><p>%1$s</p></div>', $message );

    }

    /**
     * Admin notice
     *
     * Warning when the site doesn't have a minimum required Elementor version.
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function admin_notice_minimum_elementor_version() {

        if ( isset( $_GET['activate'] ) ) unset( $_GET['activate'] );

        $message = sprintf(
            /* translators: 1: Plugin name 2: Elementor 3: Required Elementor version */
            esc_html__( '"%1$s" requires "%2$s" version %3$s or greater.', 'jfl-style-cleaner' ),
            '<strong>' . esc_html__( 'Style Cleaner for Elementor', 'jfl-style-cleaner' ) . '</strong>',
            '<strong>' . esc_html__( 'Elementor', 'jfl-style-cleaner' ) . '</strong>',
             self::MINIMUM_ELEMENTOR_VERSION
        );

        printf( '<div class="notice notice-warning is-dismissible"><p>%1$s</p></div>', $message );

    }

    /**
     * Admin notice
     *
     * Warning when the site doesn't have a minimum required PHP version.
     *
     * @since 0.0.1
     *
     * @access public
     */
    public function admin_notice_minimum_php_version() {

        if ( isset( $_GET['activate'] ) ) unset( $_GET['activate'] );

        $message = sprintf(
            /* translators: 1: Plugin name 2: PHP 3: Required PHP version */
            esc_html__( '"%1$s" requires "%2$s" version %3$s or greater.', 'jfl-style-cleaner' ),
            '<strong>' . esc_html__( 'Style Cleaner for Elementor', 'jfl-style-cleaner' ) . '</strong>',
            '<strong>' . esc_html__( 'PHP', 'jfl-style-cleaner' ) . '</strong>',
             self::MINIMUM_PHP_VERSION
        );

        printf( '<div class="notice notice-warning is-dismissible"><p>%1$s</p></div>', $message );

    }


    public function enqueue_editor_scripts() {

        wp_register_script( 'jfl-style-cleaner', plugins_url( 'assets/js/jfl-style-cleaner.js', __FILE__ ), [ 'jquery' ] );
        wp_enqueue_script('jfl-style-cleaner');

    }

}

Elementor_Test_Extension::instance();